sky/skylet/providers/lambda_cloud/node_provider.py

```diff
@@ -50,81 +52,101 @@ def __init__(self, provider_config: Dict[str, Any],
-                        TAG_RAY_USER_NODE_TYPE: 'ray_head_default',
-                        TAG_RAY_NODE_NAME: f'ray-{cluster_name}-head',
-                        TAG_RAY_LAUNCH_CONFIG: launch_hash,
-                    }
-                }
+        self.metadata = lambda_utils.Metadata(_TAG_PATH_PREFIX, cluster_name)
+        self.ssh_key_path = os.path.expanduser(auth.PRIVATE_SSH_KEY_PATH)
+        remote_ssh_key = os.path.expanduser(_REMOTE_RAY_SSH_KEY)
+        if os.path.exists(remote_ssh_key):
+            self.ssh_key_path = remote_ssh_key
+
+    def _guess_and_add_missing_tags(self, vms: Dict[str, Any]) -> None:
+        """Adds missing vms to local tag file and guesses their tags."""
+        for node in vms:
+            if self.metadata.get(node['id']) is not None:
+                pass
+            elif node['name'] == f'{self.cluster_name}-head':
+                self.metadata.set(
+                    node['id'], {
+                        'tags': {
+                            TAG_RAY_CLUSTER_NAME: self.cluster_name,
+                            TAG_RAY_NODE_STATUS: STATUS_UP_TO_DATE,
+                            TAG_RAY_NODE_KIND: NODE_KIND_HEAD,
+                            TAG_RAY_USER_NODE_TYPE: 'ray_head_default',
+                            TAG_RAY_NODE_NAME: f'ray-{self.cluster_name}-head',
+                        }
+                    })
+            elif node['name'] == f'{self.cluster_name}-worker':
+                self.metadata.set(
+                    node['id'], {
+                        'tags': {
+                            TAG_RAY_CLUSTER_NAME: self.cluster_name,
+                            TAG_RAY_NODE_STATUS: STATUS_UP_TO_DATE,
+                            TAG_RAY_NODE_KIND: NODE_KIND_WORKER,
+                            TAG_RAY_USER_NODE_TYPE: 'ray_worker_default',
+                            TAG_RAY_NODE_NAME: f'ray-{self.cluster_name}-worker',
+                        }
+                    })
 
     def _list_instances_in_cluster(self) -> Dict[str, Any]:
         """List running instances in cluster."""
         vms = self.lambda_client.list_instances()
-        return [node for node in vms if node.get('name') == self.cluster_name]
+        possible_names = [
+            f'{self.cluster_name}-head', f'{self.cluster_name}-worker'
+        ]
+        return [node for node in vms if node.get('name') in possible_names]
 
     @synchronized
     def _get_filtered_nodes(self, tag_filters: Dict[str,
                                                     str]) -> Dict[str, Any]:
 
-        def match_tags(vm):
-            vm_info = self.metadata[vm['id']]
+        def _extract_metadata(vm: Dict[str, Any]) -> Dict[str, Any]:
+            metadata = {'id': vm['id'], 'status': vm['status'], 'tags': {}}
+            instance_info = self.metadata.get(vm['id'])
+            if instance_info is not None:
+                metadata['tags'] = instance_info['tags']
```

Comment:
Seems we have changed the `__setitem__` to `set()`, should we use `set()` instead here?

Reply:
You are very right; how did the smoke tests not catch this haha

Reply:
Actually -- this `metadata` and `self.metadata` are two different things (the former is `self.cached_nodes`)

Reply:
Ahh, that makes sense!